hal/src/sercom/v2/uart/mod.rs

```diff
@@ -0,0 +1,1528 @@
+}
+impl Transmit for TxDuplex {}
+
+//=============================================================================
+// Config
+//=============================================================================
+
+/// A configurable, disabled UART peripheral
+///
+/// This `struct` represents a configurable UART peripheral in its disabled
+/// state. It is generic over the set of [`Pads`] and [`CharSize`].
+/// Upon creation, the [`Config`] takes ownership of the
+/// [`Sercom`] and resets it, returning it configured as an UART peripheral
+/// with a default configuration:
+///
+/// * [`EightBit`]
+/// * No parity
+/// * One stop bit
+/// * LSB-first
+///
+/// [`Config`] uses a builder-pattern API to configure the peripheral,
+/// culminating in a call to [`enable`], which consumes the [`Config`] and
+/// returns enabled [`Uart`]. The [`enable`] method is
+/// restricted to [`ValidConfig`]s.
+///
+/// [`enable`]: Config::enable
+pub struct Config<P, C = EightBit>
+where
+    P: ValidPads,
+    C: CharSize,
+{
+    registers: Registers<P::Sercom>,
+    pads: P,
+    chsize: PhantomData<C>,
+    freq: Hertz,
+}
+
+#[cfg(any(feature = "samd11", feature = "samd21"))]
+type Clock = pac::PM;
+
+#[cfg(feature = "min-samd51g")]
+type Clock = pac::MCLK;
+
+impl<P: ValidPads> Config<P> {
+    /// Create a new [`Config`] in the default configuration
+    ///
+    /// This function will enable the corresponding APB clock, reset the
+    /// [`Sercom`] peripheral, and return a [`Config`] in the default
+    /// configuration:
+    ///
+    /// * [`EightBit`] [`CharSize`]
+    /// * No parity
+    /// * One stop bit
+    /// * LSB-first
+    ///
+    /// [`Config`] takes ownership of the [`Sercom`] and [`Pads`].
+    ///
+    /// Users must configure GCLK manually. The `freq` parameter represents the
+    /// GCLK frequency for this [`Sercom`] instance.
+    pub fn new(clk: &Clock, mut sercom: P::Sercom, pads: P, freq: impl Into<Hertz>) -> Self {
+        sercom.enable_apb_clock(clk);
+        Self::create(sercom, pads, freq).msb_first(false)
+    }
+
+    /// Create a new [`Config`] in the default configuration
+    #[inline]
+    fn create(sercom: P::Sercom, pads: P, freq: impl Into<Hertz>) -> Self {
+        let mut registers = Registers { sercom };
+        Self::swrst(&mut registers);
+
+        let sercom = registers.sercom();
+
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        P::configure(&sercom);
+
+        #[cfg(feature = "min-samd51g")]
+        {
+            <P as Rxpo>::configure(&sercom);
+            <P as Txpo>::configure(&sercom);
+        }
+
+        EightBit::configure(&sercom);
+
+        // Enable internal clock mode
+        registers
+            .usart()
+            .ctrla
+            .modify(|_, w| w.mode().variant(MODE_A::USART_INT_CLK));
+
+        Self {
+            registers,
+            pads,
+            chsize: PhantomData,
+            freq: freq.into(),
+        }
+    }
```

Comment:
I started implementing a `Registers` type myself, and I realized a few things.

1. I think a better name for this function is `default`. Do you agree? I never did like the name `create`.
2. You could add a `.reset()` method to `Registers` and get rid of the `.swrst()` method on `Config`.
3. I decided to change the `.configure()` trait method approach on `ValidPads`, `CharSize`, etc. Instead of passing a `&RegisterBlock` or something else to the trait, I pass the trait's associated constants to a method on `Registers`. I personally like that structure more. I think it better reflects that the `Registers` type is responsible for configuring the registers, and the trait is responsible for providing the constants. It also gets rid of the ugly `<T as Trait>::method()` syntax. Instead, you get `registers.set_rxpo(P::RXPO)`.

None of this is mandatory though. I have a tendency to iterate indefinitely if left to my own devices.